File: typescript/packages/core/src/types/Chain.ts
/**
 * @param type - "evm" or "solana", extend this union as needed (e.g., "sui")
 * @param id - Chain ID, optional for EVM
 */

// Structure for native currency info (mirroring viem)
export type NativeCurrency = {
    name: string;
    symbol: string;
    decimals: number;
};

export type Chain =
    | EvmChain
    | SolanaChain
    | AptosChain
    | ChromiaChain
    | FuelChain
    | SuiChain
    | ZilliqaChain
    | CosmosChain
    | StarknetChain
    | RadixChain
    | ZetrixChain;

export type SuiChain = {
    type: "sui";
};

export type EvmChain = {
    type: "evm";
    id: number;
    nativeCurrency: NativeCurrency;
};

export type SolanaChain = {
    type: "solana";
    nativeCurrency: NativeCurrency;
};

export type AptosChain = {
    type: "aptos";
};

export type ChromiaChain = {
    type: "chromia";
};

export type FuelChain = {
    type: "fuel";
};

export type ZilliqaChain = {
    type: "zilliqa";
    id: number;
    evmId: number;
};

export type CosmosChain = {
    type: "cosmos";
};

export type StarknetChain = {
    type: "starknet";
};

export type RadixChain = {
    type: "radix";
    id: number;
};

export type ZetrixChain = {
    type: "zetrix";
};

File: python/src/goat-sdk/goat/types/chain.py
from typing import TypedDict, Literal, Union

class NativeCurrency(TypedDict):
    """Native currency type definition
    
    Args:
        name: Currency name
        symbol: Currency symbol
        decimals: Number of decimals
    """
    name: str
    symbol: str
    decimals: int

class EvmChain(TypedDict):
    """EVM chain type definition
    
    Args:
        type: Literal "evm" chain type identifier
        id: Chain ID number for EVM networks
        nativeCurrency: Native currency information
    """
    type: Literal["evm"]
    id: int
    nativeCurrency: NativeCurrency

class SolanaChain(TypedDict):
    """Solana chain type definition
    
    Args:
        type: Literal "solana" chain type identifier
        nativeCurrency: Native currency information
    """
    type: Literal["solana"]
    nativeCurrency: NativeCurrency

class AptosChain(TypedDict):
    """Aptos chain type definition
    
    Args:
        type: Literal "aptos" chain type identifier
    """
    type: Literal["aptos"]

class ChromiaChain(TypedDict):
    """Chromia chain type definition
    
    Args:
        type: Literal "chromia" chain type identifier
    """
    type: Literal["chromia"]

class MultiversXChain(TypedDict):
    """MultiversX chain type definition
    
    Args:
        type: Literal "multiversx" chain type identifier
    """
    type: Literal["multiversx"]

Chain = Union[EvmChain, SolanaChain, AptosChain, ChromiaChain, MultiversXChain]